src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -1,46 +1,100 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Options.Providers;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Options
 {
     [Export(typeof(IGlobalOptionService)), Shared]
     internal class GlobalOptionService : IGlobalOptionService
     {
-        private readonly Lazy<HashSet<IOption>> _options;
+        private readonly Lazy<ImmutableHashSet<IOption>> _lazyAllOptions;
         private readonly ImmutableArray<Lazy<IOptionPersister>> _optionSerializers;
+        private readonly ImmutableDictionary<string, ImmutableArray<Lazy<IOptionProvider, LanguageMetadata>>> _languageToLazyProvidersMap;
+        private readonly Dictionary<string, ImmutableHashSet<IOption>> _languageToSerializableOptionsMap;
+        private readonly HashSet<string> _forceComputedLanguages;
+        private readonly ImmutableHashSet<Workspace>.Builder _registeredWorkspacesBuilder;
```

Comment:
>_registeredWorkspacesBuilder [](start = 61, length = 28)

Wouldn't `ImmutableArray<Workspace>` be good enough? There won't be that many workspaces.

Reply:
Also, I'd call this just `_registeredWorkspaces`. 

---
In reply to: [362685132](https://github.com/dotnet/roslyn/pull/40289#discussion_r362685132) [](ancestors = 362685132)

Reply:
Couple of issues with making it an immutable array:
1. We are adding/removing workspaces with Register/Unregister methods, so it would be kind of awkward to do these operations on an array.
2. The field can no longer be readonly.

Reply:
Re [1] that shouldn't be a problem, ImmutableArray.Add() returns a new array that you can assign back to the field.
Re [2] readonly-ness of fields that hold on a mutable state isn't really that much useful.

Reply:
Sounds good, done.